pkg/operatorstatus/builder.go

```diff
@@ -0,0 +1,190 @@
+		Status:             status,
+		Message:            message,
+		LastTransitionTime: metav1.NewTime(b.clock.Now()),
+	}
+
+	b.setCondition(condition)
+
+	return b
+}
+
+// WithVersion adds the specific version into the status.
+func (b *Builder) WithVersion(name, version string) *Builder {
+	b.init()
+
+	existing := b.findVersion(name)
+	if existing != nil {
+		existing.Version = version
+		return b
+	}
+
+	ov := configv1.OperandVersion{
+		Name:    name,
+		Version: version,
+	}
+	b.status.Versions = append(b.status.Versions, ov)
+
+	return b
+}
+
+// WithRelatedObject adds the reference specified to the RelatedObjects list.
+func (b *Builder) WithRelatedObject(group, resource, namespace, name string) *Builder {
+	b.init()
+
+	reference := configv1.ObjectReference{
+		Group:     group,
+		Resource:  resource,
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	b.setRelatedObject(reference)
+
+	return b
+}
+
+func (b *Builder) init() {
+	if b.status == nil {
+		b.status = &configv1.ClusterOperatorStatus{
+			Conditions:     []configv1.ClusterOperatorStatusCondition{},
+			Versions:       []configv1.OperandVersion{},
+			RelatedObjects: []configv1.ObjectReference{},
+		}
+	}
+}
+
+func (b *Builder) findCondition(conditionType configv1.ClusterStatusConditionType) *configv1.ClusterOperatorStatusCondition {
+	for i := range b.status.Conditions {
+		if b.status.Conditions[i].Type == conditionType {
+			return &b.status.Conditions[i]
+		}
+	}
+
+	return nil
+}
+
+func (b *Builder) setCondition(condition *configv1.ClusterOperatorStatusCondition) {
+	existing := b.findCondition(condition.Type)
+	if existing == nil {
+		b.status.Conditions = append(b.status.Conditions, *condition)
+		return
+	}
+
+	existing.Reason = condition.Reason
+	existing.Message = condition.Message
+
+	if existing.Status != condition.Status {
+		existing.Status = condition.Status
+		existing.LastTransitionTime = condition.LastTransitionTime
+	}
+}
+
+func (b *Builder) findVersion(name string) *configv1.OperandVersion {
+	for i := range b.status.Versions {
+		if b.status.Versions[i].Name == name {
+			return &b.status.Versions[i]
+		}
+	}
+
+	return nil
+}
+
+func (b *Builder) setRelatedObject(reference configv1.ObjectReference) {
+	for i := range b.status.RelatedObjects {
+		if reflect.DeepEqual(b.status.RelatedObjects[i], reference) {
+			return
+		}
+	}
+
+	b.status.RelatedObjects = append(b.status.RelatedObjects, reference)
+}
+
+// WithMarketplaceRelatedObjects populates RelatedObjects that are related to Marketplace.
+// RelatedObjects are consumed by https://github.com/openshift/must-gather
+func (b *Builder) WithMarketplaceRelatedObjects(namespace string) *Builder {
+	// Add the operator's namespace which will result in core resources being gathered
+	b.WithRelatedObject("", "namespaces", "", namespace)
+
+	// Add the non-core resources we care about
+	b.WithRelatedObject(v1.SchemeGroupVersion.Group, v1.OperatorSourceKind, namespace, "").
```

Comment:
Please do a test with a private `OperatorSource` and ensure that other core resources created are being added. I remember testing this very early on but we should confirm this.

Reply:
Will do.